flowmachine/flowmachine/features/subscriber/contact_reciprocal.py

```diff
@@ -0,0 +1,388 @@
+            COALESCE(O.events_out, 0) AS events_out,
+            COALESCE(I.proportion_in, 0) AS proportion_in,
+            COALESCE(O.proportion_out, 0) AS proportion_out,
+            CASE
+                WHEN I.events_in IS NULL OR O.events_out IS NULL THEN FALSE
+                ELSE TRUE
+            END AS reciprocal
+        FROM (
+            SELECT
+                subscriber,
+                msisdn_counterpart,
+                events AS events_in,
+                proportion AS proportion_in
+            FROM ({self.contact_in_query.get_query()}) C
+        ) I
+        FULL OUTER JOIN (
+            SELECT
+                subscriber,
+                msisdn_counterpart,
+                events AS events_out,
+                proportion AS proportion_out
+            FROM ({self.contact_out_query.get_query()}) C
+        ) O
+        ON
+            I.subscriber = O.subscriber AND
+            I.msisdn_counterpart = O.msisdn_counterpart
+        """
+
+        return sql
+
+
+class ProportionContactReciprocal(SubscriberFeature):
+    """
+    This class calculates the proportion of reciprocal contacts a subscriber has.
+
+    A reciprocal contact is a contact who has initiated contact and who also
+    has been the counterpart of an initatiated contact by the subscriber.
+
+    Parameters
+    ----------
+    start, stop : str
+         iso-format start and stop datetimes
+    hours : 2-tuple of floats, default 'all'
+        Restrict the analysis to only a certain set
+        of hours within each day.
+    subscriber_subset : str, list, flowmachine.core.Query, flowmachine.core.Table, default None
+        If provided, string or list of string which are msisdn or imeis to limit
+        results to; or, a query or table which has a column with a name matching
+        subscriber_identifier (typically, msisdn), to limit results to.
+    exclude_self_calls : bool, default True
+        Set to false to *include* calls a subscriber made to themself
+    tables : str or list of strings, default 'all'
+        Can be a string of a single table (with the schema)
+        or a list of these. The keyword all is to select all
+        subscriber tables
+
+    Example
+    -------
+
+    >> s = ProportionContactReciprocal('2016-01-01', '2016-01-08')
+    >> s.get_dataframe()
+
+          subscriber       value
+    9vXy462Ej8V1kpWl         0.0
+    Q4mwVxpBOo7X2lb9         0.0
+    5jLW0EWeoyg6NQo3         0.0
+    QEoRM9vlkV18N4ZY         0.0
+    a76Ajyb9dmEYNd8L         0.0
+                 ...         ...
+    """
+
+    def __init__(
+        self,
+        start,
+        stop,
+        *,
+        hours="all",
+        tables="all",
+        exclude_self_calls=True,
+        subscriber_subset=None,
+    ):
+        self.start = start
+        self.stop = stop
+        self.hours = hours
+        self.exclude_self_calls = exclude_self_calls
+        self.tables = tables
+
+        self.contact_reciprocal_query = ContactReciprocal(
```

Comment:
Couldn't the class just take a `ContactReciprocal` object as the argument?

Reply:
Good point. But my intention was to homogenize the call for feature extraction such that all of them have roughly the same parameters, that is, `start`, `stop`, `hours`, `tables`, etc. I find this approach is somewhat cleaner than having too many variations across feature classes. In fact, I would ideally like to see the init parameters for query classes defined in a parent class.

Reply:
My rationale here is that bubbling up the subquery is slightly better from an API call & caching perspective, and there's not really any point in storing the params and passing them straight on down.

Reply:
Does it mean, that you rather using `ContactReciprocal` as an argument or that I do not save pointers to the initialized params and rather just save a pointer to `ContactReciprocal`?

If we do decide to have `ContactReciprocal` as one of the arguments, could we perhaps allow the user to either pass all the usual ones or `ContactReciprocal`? Or is this bad coding?

I generally find convenient to be able to reference to attributes of the query class such as start, stop, tables, etc. But I don't know if that's the design choice we would like to make.

Reply:
`ContactReciprocal` as only argument. I hate cluttered signatures.